rust/kernel/user_ptr.rs

```diff
@@ -144,27 +150,40 @@ impl UserSlicePtrReader {
     /// Returns `EFAULT` if the byte slice is bigger than the remaining size
     /// of the user slice or if the address does not currently point to mapped,
     /// readable memory.
-    pub fn read(&mut self, data: &mut [u8]) -> error::KernelResult<()> {
-        if data.len() > self.1 || data.len() > u32::MAX as usize {
+    pub fn read_slice(&mut self, data: &mut [u8]) -> KernelResult<()> {
+        // SAFETY: The output buffer is valid as it's coming from a live reference.
+        unsafe { self.read_raw(data.as_mut_ptr(), data.len()) }
+    }
+
+    /// Reads raw data from the user slice into a raw kernel buffer.
+    ///
+    /// # Safety
+    ///
+    /// The output buffer must be valid.
+    pub unsafe fn read_raw(&mut self, out: *mut u8, len: usize) -> KernelResult<()> {
+        if len > self.1 || len > u32::MAX as usize {
             return Err(error::Error::EFAULT);
         }
-        let res = unsafe {
-            rust_helper_copy_from_user(
-                data.as_mut_ptr() as *mut c_types::c_void,
-                self.0,
-                data.len() as _,
-            )
-        };
+        let res = rust_helper_copy_from_user(out as _, self.0, len as _);
         if res != 0 {
             return Err(error::Error::EFAULT);
         }
         // Since this is not a pointer to a valid object in our program,
         // we cannot use `add`, which has C-style rules for defined
         // behavior.
-        self.0 = self.0.wrapping_add(data.len());
-        self.1 -= data.len();
+        self.0 = self.0.wrapping_add(len);
+        self.1 -= len;
         Ok(())
     }
+
+    /// Reads the contents of a plain old data (POD) type from the user slice.
+    pub fn read<T: Copy>(&mut self) -> KernelResult<T> {
```

Comment:
`Copy` isn't necessarily the exact same thing as "safe to instantiate from arbitrary bytes" -- for example, `bool` is copy, but it's UB if it exists with values besides `0` or `1`.

Reply:
That's unfortunate. Do you know of a way to filter structs such that fields only contain say `u32`s and `u64`s (or whatever else we vet as ok)? Or is making this unsafe the way to go?

Reply:
That's basically what https://github.com/rust-lang/rfcs/pull/2981 is trying to define. I think maybe the best thing would be to have callers just get slices and do their own transmute for now, and then we can add a safe API as soon as safe-transmute is supported? WDYT.

Reply:
There are two aspects that I don't like about getting slices and transmuting:
1. The vast majority of structs being read/written are POD without bools, so I believe a lot of duplicate would be written for the same purpose. People who do have `bool`s in their structs would likely just copy/paste the code and therefore introduce UBs controlled by userspace (like I was about to do here).
2. Performance: some drivers, like binder, are performance-sensitive. If we require them to copy data to slices first (potentially allocating memory), then we may make it prohibitively expensive to port some code to rust.
 
If we provide an unsafe `read<T: Copy>` explaining that the caller must ensure that the struct fields all have acceptable types, we will avoid repeating code in drivers and hopefully people will learn about the pitfalls when reading the Safety section.

WDYT?

Reply:
I can live with it :-) I suppose down the line we can always change the bound on `read<T: TransmuteSafe>` and make it `safe` and that's fine :-)

Reply:
Can we make it fail to compile via `static_assert!` or similar?

Reply:
Do you know of a way to check for a type with a `static_assert`? We could use the size, but that would exclude other acceptable types...

Reply:
Yeah, that was my C++ brain talking late in the night, never mind.

Reply:
After making this change, I was cringing while adding unsafe blocks to a lot of places where I call this function. So I thought of a better design: define an unsafe trait that declares a type as safe to from byte slices, and constrain T to that. It has the advantage that the safety requirements are enforced on the type, not when `read` is called; the former only happens once, while the latter happens repeatedly.

We can then retire the trait once Rust defines one.

The changes are here: https://github.com/wedsonaf/linux/commit/22372040c0e5adfbb678dc5ee2bea6a2e81d7073 -- I don't have write access to this PR anymore.

Reply:
Pushed.

Reply:
Yeah, if there are many callers, having a custom unsafe trait looks much better!